src/ros_tcp_endpoint/tcp_sender.py

```diff
@@ -86,4 +88,32 @@ def send_unity_service(self, topic, service_class, request):
             return response
         except Exception as e:
             rospy.loginfo("Exception {}".format(e))
-            
\ No newline at end of file
+ 
+    def sender_loop(self):
+        s = None
+        sockettimeout = 0
+        
+        while True:
+            item = self.queue.get()
+            
+            if time.time() > sockettimeout:
+                if s != None:
+                    s.close()
+
+                try:
+                    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+                    s.settimeout(self.timeout)
+                    s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+                    s.connect((self.unity_ip, self.unity_port))
+                    s.sendall(struct.pack('<f', self.timeout))
+                except Exception as e:
+                    rospy.loginfo("Exception on Connect {}".format(e))
+                    sockettimeout = 0
+            
+            try:
+                s.sendall(item)
+                # on the endpoint, make a new connection after 3/4ths of the timeout time. Trying to avoid
+                # a situation where Unity has stopped listening but the endpoint is still trying to send
+                sockettimeout = time.time() + self.timeout * 0.75
```

Comment:
`0.75` seems very much like a magic number. Rather than trying to finesse around this particular case, could you add explicit handling for the exception that's thrown when the endpoint tries to send a message that Unity fails to receive?  What is the desired behavior in that case?

Reply:
Yeah, clearly a magic number :-)
The problem is that when the endpoint tries to send a message but Unity isn't listening, it blocks for about 15 seconds before actually throwing the exception. I'm open to suggestions about how to avoid this situation.

Reply:
Hmmm well if the application is blocking on socket connection, then I can think of a couple things to try: The first is to make the connection attempt asynchronous; as a general rule anything with a long timeout shouldn't happen on the main thread. If it's already in a thread and the blocking is problematic for another reason, add a small per-attempt timeout which is based on how long you'd expect it to take in the optimal case, add a small sleep between attempts, and either use that 15 seconds as a "master timeout" to limit the amount of  retries, or explicitly set the number of connection attempts allowed.

Another sort of tangential thing is that I don't think we're using the python `logging` module in this package, but that would be really helpful here for putting spammy log messages from short connection timeouts into a debug-level message, and only printing a warning or error severity message when all the retries have been exhausted.

Reply:
Yeah, the sender_loop function here is already its own thread, so the only thing that's being blocked is future sends.

Ooh, actually, I just realized we're already setting a timeout (line 105): the only problem is it's way too long.